docs/architecture/adr-103-proto-versioning.md

```diff
@@ -0,0 +1,140 @@
+# ADR 103: Protobuf definition versioning
+
+## Changelog
+
+- 2023-04-27: First draft (@thanethomson)
+
+## Status
+
+Accepted
+
+## Context
+
+In CometBFT v0.34 through v0.38, the Protocol Buffers definitions in the `proto`
+folder are organized in a relatively flat structure and evolve over time in
+major releases of CometBFT.
+
+For integrators who may want to interact with nodes running different versions
+of CometBFT, this means that, when consuming those Protobuf definitions, they
+need to clone the CometBFT repository and check out the definitions for each and
+every major version they want to support. It also means that they need to
+manually diff the types to get a precise sense of what changed between CometBFT
+releases.
+
+We hypothesize that, if we had to version our Protobuf definitions according to
+[Buf's style guide][buf-style], we could simplify integration efforts with
+CometBFT in the long run. Adopting such a practice would also introduce a new
+level of rigour into the way CometBFT Protobuf definitions are managed, allowing
+for better expectation management with integrators. It would also allow the core
+team to make better use of Protobuf-related tooling (such as Buf) in enforcing
+standards, conventions and versioning practices in an automated way.
+
+Some context is captured in [\#95] and the issues/PRs to which it links.
+
+## Alternative Approaches
+
+1. The main alternative here is to do nothing and keep our existing Protobuf
+   definition approach, which suffers from the issues captured in the
+   [Context](#context) section.
+2. Version our Protobuf definitions, as per this ADR. The primary drawback of
+   introducing versioning is that it represents a substantial breaking change.
+   Within this approach, there are two alternative policies that @mzabaluev has
+   proposed:
+   1. **Conservative**, recommended by Buf: Do not create new versions when
+      making non-breaking changes to Protobuf message definitions, such as
+      adding a new field. The upside to this approach is that non-breaking
+      changes can be made in non-breaking releases of CometBFT, but the downside
+      is that some programming languages, such as Rust, cannot generate code in
```

Comment:
> is that some programming languages

Isn't it the combination of the programming language and the protobuf code generation libraries available for that language?